### PR TITLE
Modular decom

### LINKED ISSUE #27: Modular Decommutation Logic
This change will provide much better support for non-binary packets.  
In particular, the raw data as received can be logged rather than a version converted first to binary.

### PR DESCRIPTION
**Pull request by ryanmelt**
closes #27

### CODE REVIEW

File: openc3/lib/openc3/packets/packet.rb [hunk 1]
```diff
@@ -1032,6 +1112,22 @@ def self.from_json(hash)
       packet
     end
 
+    def decom
+      # Read all the RAW at once because this could be optimized by the accessor
+      json_hash = read_items(@sorted_items)
```
Comment (jmthomas) [inline openc3/lib/openc3/packets/packet.rb:1117]: So read_items initializes the hash and then we call read_item again but pass given_raw to really populate it? I'm confused about the 2 pass system here.

File: openc3/lib/openc3/packets/structure_item.rb [hunk 2]
```diff
@@ -297,13 +308,16 @@ def self.from_json(hash)
       endianness = hash['endianness'] ? hash['endianness'].intern : nil
       data_type = hash['data_type'] ? hash['data_type'].intern : nil
       overflow = hash['overflow'] ? hash['overflow'].intern : nil
-      StructureItem.new(hash['name'], hash['bit_offset'], hash['bit_size'], data_type,
+      si = StructureItem.new(hash['name'], hash['bit_offset'], hash['bit_size'], data_type,
         endianness, hash['array_size'], overflow)
+      si.key = hash['key'] || hash['name']
```
Comment (jmthomas) [inline openc3/lib/openc3/packets/structure_item.rb:313]: You didn't want to add key to the constructor?

File: openc3/spec/packets/structure_item_spec.rb [hunk 3]
```diff
@@ -221,9 +221,10 @@ module OpenC3
       it "creates a Hash" do
         item = StructureItem.new("test", 0, 8, :UINT, :BIG_ENDIAN, 16)
         hash = item.as_json(:allow_nan => true)
-        expect(hash.keys.length).to eql 7
+        expect(hash.keys.length).to eql 8
         expect(hash.keys).to include('name', 'bit_offset', 'bit_size', 'data_type', 'endianness', 'array_size', 'overflow')
         expect(hash["name"]).to eql "TEST"
+        expect(hash["key"]).to eql "test"
```
Comment (jmthomas) [inline openc3/spec/packets/structure_item_spec.rb:227]: After all those changes this is the only unit test change?

**Review by jmthomas — APPROVED** [review]
I need an example in the demo of how this works on the user side. Don't need every accessor but at least JSON. Also you probably need more unit tests.

File: openc3-init/plugins/packages/openc3-accessor-test/targets/ACCESS/target.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+IGNORE_PARAMETER ID_ITEM
+
+CMD_UNIQUE_ID_MODE
+TLM_UNIQUE_ID_MODE
```
Comment (jmthomas) [inline openc3-init/plugins/packages/openc3-accessor-test/targets/ACCESS/target.txt:4]: Are these required for the new accessor modes? Or if you have the exact same ID_ITEM KEY then no?

File: openc3/lib/openc3/packets/structure.rb [hunk 5]
```diff
@@ -200,11 +200,8 @@ def rename_item(item_name, new_item_name)
     #   dependant on the item_class but by default see StructureItem.
     # @return [StrutureItem] The struture item defined
     def define_item(name, bit_offset, bit_size, data_type, array_size = nil, endianness = @default_endianness, overflow = :ERROR)
-      # Handle case-insensitive naming
-      name_upcase = name.upcase
-
       # Create the item
-      item = @item_class.new(name_upcase, bit_offset, bit_size, data_type, endianness, array_size, overflow)
+      item = @item_class.new(name, bit_offset, bit_size, data_type, endianness, array_size, overflow)
```
Comment (jmthomas) [inline openc3/lib/openc3/packets/structure.rb:204]: Does this have other repercussions?

File: openc3-init/plugins/packages/openc3-accessor-test/targets/ACCESS/target.txt (hunk 4 above)
Comment (ryanmelt) [reply openc3-init/plugins/packages/openc3-accessor-test/targets/ACCESS/target.txt:4]: They are not required if all the packets have the same Accessor, ID_ITEMS, and keys

**Comment by jmthomas** [comment]
I love the new example project with the example of all the different ways you can access the data. This will make it much easier to use for end users. Nicely done!

**Comment by ryanmelt** [comment]
Tests pass locally.  Merging for now.

**Pull request merged**